src/worker/handlers.js

```diff
@@ -0,0 +1,118 @@
+import Logger from '../logger/logger';
+import { getToken, setToken } from './storage';
+import { GATEWAY_URL, INDEX_URL, WHITELISTED_URLS } from './config';
+
+// Logger
+const logger = new Logger();
+// Token
+let token = null;
+
+/**
+ * Is User authorized
+ *
+ * @returns {Boolean}
+ */
+function isAuthorized() {
+    return token;
+}
+
+/**
+ * Is request whitelisted
+ *
+ * @param {any} request
+ * @returns {Boolean}
+ */
+function isWhitelisted(request) {
+    return WHITELISTED_URLS.find(item => request.url.indexOf(item) !== -1);
+}
+
+/**
+ * Is Gateway request
+ *
+ * @param {any} request
+ * @returns {Boolean}
+ */
+function isGateway(request) {
+    return request.url.indexOf(GATEWAY_URL) !== -1;
+}
+
+/**
+ * Handles Activate Event
+ * 
+ * @export
+ * @param {any} event 
+ */
+export function handleActivate(event) {
+    event.waitUntil(
+        getToken().then((storedToken) => {
+            token = storedToken;
+            logger.log(`[Service worker] Activated with token: ${token}`);
+        }),
+    );
+}
+
+/**
+ * Handles Message Event
+ * 
+ * @export
+ * @param {any} event 
+ */
+export function handleMessage(event) {
+  
+    if (event.data && event.data.type) {
+        logger.log(`[Service worker] Message recieved: ${event.data}`);
+
+        const client = event.source;
+        const { type, payload } = event.data;
+
+        switch (type) {
+            case 'SET_TOKEN':
+                token = payload.token;
+                setToken(token).then(() => {
+                    client.postMessage({ type: 'TOKEN_SET' });
+                });
+                break;
+            default:
+
+        }
+
+    }
+    else {
+        logger.log('[Service worker] Un-catched message:');
+    }
+}
+
+/**
+ * Handles Fetch Event
+ * 
+ * @export
+ * @param {any} event 
+ */
+export function handleFetch(event) {
+    logger.log(`[Service worker] Fetch event: ${event.request.url}`);
+
+    if (isAuthorized() || isWhitelisted(event.request)) {
+        if (isGateway(event.request)) {
```

Comment:
I don't understand this clause - is it possible for somethng to be whitelisted _and_ gateway? why would we redirect is authorized and gateway?

sorry if I'm being dumb